cpp/src/arrow/io/io-file-test.cc

```diff
@@ -614,6 +614,31 @@ TEST_F(TestMemoryMappedFile, WriteRead) {
   }
 }
 
+TEST_F(TestMemoryMappedFile, WriteResizeRead) {
+  const int64_t buffer_size = 1024;
+  const int reps = 5;
+  std::vector<std::vector<uint8_t>> buffers(reps);
+  for (auto& b : buffers) {
+    b.resize(buffer_size);
+    test::random_bytes(buffer_size, 0, b.data());
+  }
+
+  std::string path = "io-memory-map-write-read-test";
+  std::shared_ptr<MemoryMappedFile> result;
+  ASSERT_OK(InitMemoryMap(buffer_size, path, &result));
+
+  int64_t position = 0;
+  std::shared_ptr<Buffer> out_buffer;
+  for (int i = 0; i < reps; ++i) {
+    ASSERT_OK(result->Write(buffers[i].data(), buffer_size));
```

Comment:
Per comments above, these unit tests need to exercise some of the edge cases:

* If you write to the exactly the end of the file, then write more bytes (triggering the resizing logic in my review above), will you overwrite a byte from the previous write? I fear you may
* Check file size after resize to assert that it has doubled

Additionally, can you add a test about creating a 0-length memory map and then writing data to it, to ensure that that edge case is not broken

Reply:
Done in `WriteThenShrinkToHaldThenWrite`